### PR TITLE
sparse reg op on feature map based on KL

### PR DESCRIPTION
**Pull request by winstywang**

### CODE REVIEW

**Comment by piiswrong** [comment]
LGTM
side notes: do we support multiple loss with weights currently?

Also, could you add equations to comment?

**Comment by tqchen** [comment]
@winstywang please feel free to merge when you think is ok

**Comment by winstywang** [comment]
sorry for the dalay. I am out of town these days. Actually I implemented the sec 11 of the notes here: https://www.cs.toronto.edu/~hinton/absps/guideTR.pdf It is potentially useful with sigmoid activation. Please help me review it before merge.

File: src/operator/sparse_reg-inl.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+#include <utility>
+#include "./mshadow_op.h"
+#include "./operator_common.h"
+
+namespace mxnet {
+namespace op {
+
+namespace sparsereg {
+enum SparseRegOpInputs {kData};
+enum SparseRegOpOutputs {kOut};
```
Comment (tqchen) [inline src/operator/sparse_reg-inl.h]: Consider rename as ProbKLSparseReg, as the objective seems to be specific to KL

File: src/operator/sparse_reg-inl.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    DMLC_DECLARE_FIELD(penalty).set_default(0.001)
+    .describe("The tradeoff parameter for the sparseness penalty");
+    DMLC_DECLARE_FIELD(momentum).set_default(0.9)
+    .set_range(0, 1)
+    .describe("The momentum for running average");
+  }
+};  // struct SparseRegParam
+
+template<typename xpu>
+class SparseRegOp : public Operator {
```
Comment (tqchen) [inline src/operator/sparse_reg-inl.h]: Add comments about its behavior, as the regularizer does not output the value of regularization function, but original data.

Comment (tqchen) [inline src/operator/sparse_reg-inl.h]: This is like another special operator class. 
- Backward is not the gradient on output wrt. input.

Currently only XXXOutput operators have this kind of behavior for output loss functions. If we are going to add this as special class, think about naming convention. Or add it as a true operator, output regualrization value and backprop the gradient

Comment (winstywang) [inline src/operator/sparse_reg-inl.h]: It is a kind of regularization on output feature map. How about rename it to ProbKLOutputReg? @tqchen

**Comment by winstywang** [comment]
@tqchen Could you please confirm this naming? Then I can finish it asap.

File: src/operator/sparse_reg-inl.h (hunk 2 above)
Comment (tqchen) [inline src/operator/sparse_reg-inl.h]: Since from what you proposed, the feature map will continue to be used. How about
```
IdentityAttachKLSparseReg
```

**Comment by winstywang** [comment]
@tqchen fixed as requested. please see whether it is better

**Comment by tqchen** [comment]
seems you deleted the ops, i cannot see them in diff

**Comment by winstywang** [comment]
it should be OK now. I forgot to add them after renaming @tqchen

File: src/operator/identity_attach_KL_sparse_reg-inl.h [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(in_data.size(), 1);
+    CHECK_EQ(out_data.size(), 1);
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2> data = in_data[sparsereg::kData].FlatTo2D<xpu, real_t>(s);
+    Tensor<xpu, 2> out = out_data[sparsereg::kOut].FlatTo2D<xpu, real_t>(s);
+    out = F<mshadow_op::identity>(data);
```
Comment (tqchen) [inline src/operator/identity_attach_KL_sparse_reg-inl.h]: Use ```Assign( out, req[kData], F<identity>(data))```, so it skips the opwhen the request is kWriteInplace

File: src/operator/identity_attach_KL_sparse_reg-inl.h [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                        const std::vector<TBlob> &out_grad,
+                        const std::vector<TBlob> &in_data,
+                        const std::vector<TBlob> &out_data,
+                        const std::vector<OpReqType> &req,
+                        const std::vector<TBlob> &in_grad,
+                        const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2> grad_in = in_grad[sparsereg::kData].FlatTo2D<xpu, real_t>(s);
```
Comment (tqchen) [inline src/operator/identity_attach_KL_sparse_reg-inl.h]: Does this work with arbitrary shape? or do we have specific shape requirements? FlatTo2D shape(a,b,c) to(a*b, c), make sure if this is what was wanted

File: src/operator/identity_attach_KL_sparse_reg-inl.h [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  OperatorProperty* Copy() const override {
+    auto ptr = new IdentityAttachKLSparseRegProp();
+    ptr->param_ = param_;
+    return ptr;
+  }
+
+  std::string TypeString() const override {
+    return "KLSparseReg";
+  }
+
```
Comment (tqchen) [inline src/operator/identity_attach_KL_sparse_reg-inl.h]: Add DeclareBackwardDependency to depend on out_grad and in_data

File: src/operator/identity_attach_KL_sparse_reg-inl.h [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  OperatorProperty* Copy() const override {
+    auto ptr = new IdentityAttachKLSparseRegProp();
+    ptr->param_ = param_;
+    return ptr;
+  }
+
+  std::string TypeString() const override {
+    return "KLSparseReg";
```
Comment (tqchen) [inline src/operator/identity_attach_KL_sparse_reg-inl.h]: IdentityAttachKLSparseRegProp

**Comment by tqchen** [comment]
a few more comments

File: src/operator/identity_attach_KL_sparse_reg-inl.h (hunk 4 above)
Comment (winstywang) [inline src/operator/identity_attach_KL_sparse_reg-inl.h]: It is elementwise, so it should be OK

**Comment by winstywang** [comment]
all fixed

File: src/operator/identity_attach_KL_sparse_reg-inl.h (hunk 4 above)
Comment (tqchen) [inline src/operator/identity_attach_KL_sparse_reg-inl.h]: What is the sum_all_except_dim<1> for ? The average

Comment (winstywang) [inline src/operator/identity_attach_KL_sparse_reg-inl.h]: the average activation over batch
